src/library_webgpu.js

```diff
@@ -1212,60 +1258,133 @@ var LibraryWebGPU = {
+  },
+
   // wgpuBuffer
 
-  wgpuBufferSetSubData: function(bufferId, {{{ defineI64Param('start') }}}, {{{ defineI64Param('count') }}}, data) {
-    {{{ receiveI64ParamAsI32s('start') }}}
-    {{{ receiveI64ParamAsI32s('count') }}}
-    var buffer = WebGPU.mgrBuffer.get(bufferId);
-    var start = {{{ gpu.makeU64ToNumber('start_low', 'start_high') }}};
-    var count = {{{ gpu.makeU64ToNumber('count_low', 'count_high') }}};
-    buffer["setSubData"](start, HEAPU8, data, count);
-  },
-
-  wgpuBufferMapReadAsync: function(bufferId, callback, userdata) {
-    var bufferEntry = WebGPU.mgrBuffer.objects[bufferId];
-    bufferEntry.mapped = 'write';
-    var buffer = bufferEntry.object;
-
-    buffer["mapReadAsync"]().then(function(mapped) {
-      var WEBGPU_BUFFER_MAP_ASYNC_STATUS_SUCCESS = 0;
-      var data = _malloc(mapped.byteLength);
-      HEAPU8.set(new Uint8Array(mapped), data);
-      var dataLength_high = (mapped.byteLength / 0x100000000) | 0;
-      var dataLength_low = mapped.byteLength | 0;
-      // WGPUBufferMapAsyncStatus status, const void* data, uint64_t dataLength, void* userdata
-      dynCall('viiji', callback, [WEBGPU_BUFFER_MAP_ASYNC_STATUS_SUCCESS, data, {{{ sendI64Argument('dataLength_low', 'dataLength_high') }}}, userdata]);
-    }, function() {
-      // TODO(kainino0x): Figure out how to pick other error status values.
-      var WEBGPU_BUFFER_MAP_ASYNC_STATUS_ERROR = 1;
-      dynCall('viiji', callback, [WEBGPU_BUFFER_MAP_ASYNC_STATUS_ERROR, 0, 0, 0, userdata]);
+  wgpuBufferGetConstMappedRange: function(bufferId, offset, size) {
+    var bufferWrapper = WebGPU.mgrBuffer.objects[bufferId];
+
+    var mapped;
+    try {
+      mapped = bufferWrapper.object["getMappedRange"](offset, size);
+    } catch (ex) {
+      // TODO(kainino0x): Somehow inject a validation error?
+      return 0;
+    }
+
+    var data = _malloc(mapped.byteLength);
+    HEAPU8.set(new Uint8Array(mapped), data);
+    bufferWrapper.onUnmap.push(function() {
+      _free(data);
     });
+    return data;
   },
 
-  wgpuBufferMapWriteAsync: function(bufferId, callback, userdata) {
+  wgpuBufferGetMappedRange: function(bufferId, offset, size) {
     var bufferWrapper = WebGPU.mgrBuffer.objects[bufferId];
+
+    if (bufferWrapper.mapMode !== 2 /* WGPUMapMode_Write */) {
+#if ASSERTIONS
+      abort("GetMappedRange called, but buffer not mapped for writing");
+#endif
+      // TODO(kainino0x): Somehow inject a validation error?
+      return 0;
+    }
+
+    var mapped;
+    try {
+      mapped = bufferWrapper.object["getMappedRange"](offset, size);
+    } catch (ex) {
+      // TODO(kainino0x): Somehow inject a validation error?
+      return 0;
+    }
+
+    var data = _malloc(mapped.byteLength);
+    HEAPU8.fill(0, data, mapped.byteLength);
+    bufferWrapper.onUnmap.push(function() {
+      new Uint8Array(mapped).set(HEAPU8.subarray(data, data + mapped.byteLength));
+      _free(data);
+    });
+    return data;
```

Comment:
Oh, these two functions `wgpuBufferGetConstMappedRange` and `wgpuBufferGetMappedRange` are horrible 🤮 😞 

I suppose mapping a GPU buffer directly to wasm heap address space will be a hard no-go? (has that been discussed with Wasm folks? CC @lukewagner @dschuff  )

Assuming it is not possible, I wonder if we should consider changing up the API completely. Given that there is only asynchronous mapping support, we should make this a callback-based API where a cb function gets called when the map happens, and the cb function will receive a handle ID, and then have dedicated functions
```js
void wgpuMemset8ToMappedBuffer(uint destinationMappedBufferHandleId, uint destinationOffset, uint destinationSize, uchar fillValue);
void wgpuMemset16ToMappedBuffer(uint destinationMappedBufferHandleId, uint destinationOffset, uint destinationSize, ushort fillValue);
void wgpuMemset32ToMappedBuffer(uint destinationMappedBufferHandleId, uint destinationOffset, uint destinationSize, uint fillValue);
void wgpuMemcpyToMappedBuffer(uint destinationMappedBufferHandleId, uint destinationOffset, uint destinationSize, void *sourcePtr);
void wgpuMemcpyFromMappedBuffer(void *destinationPtr, uint destinationOffset, uint destinationSize, uint sourceMappedBufferHandleId);
```
This way the caller will be in control whether there will need to be malloc/free in effect, and avoid any redundant memsets and memcpys.

Reply:
We had some serious conversations about mapping memory directly into WASM heap address space, and we think it's definitely possible (it can be emulated by copies anytime the OS primitives aren't there to make it real). However we haven't talked about it much in... 3 years, at this point. Tracked here https://github.com/gpuweb/gpuweb/issues/747

The memset/memcpy idea is interesting. I'll copy this to a webgpu-headers issue.

Reply:
BTW, we do have `GPUQueue.writeBuffer` which just takes a piece of data and writes it into a buffer. It's intended to be better for a lot of usecases especially if there are extra copies in mapping.